### PR TITLE
Simplified workflow for camera serial connection and hardware trigger…

### PR DESCRIPTION
**Pull request by kheki4**
…ing. Also fixed several minor bugs related to stereo camera settings dialog and camera connection. Added hardware triggering support also for Arduino Uno microcontrollers.

### CODE REVIEW

File: src/devices/stereoCamera.cpp [hunk 1]
```diff
@@ -1472,32 +1327,59 @@ bool StereoCamera::setImageROIoffsetYEmu(int offsetY) {
...
+
+void StereoCamera::safelyCloseCameras() {
+    if(!cameras.IsOpen())
+        return;
+
+    if(cameras.IsGrabbing())
+        cameras.StopGrabbing();
+    cameras.Close();
+
+    if(cameraImageEventHandler) {
```
Comment (sasfog) [inline src/devices/stereoCamera.cpp:1358]: Event Handlers registered with delete_cleanup cleanup procedure are automatically cleaned up after deregistering, so there should be no need for manual deletion (cameraImageEventHandler->DestroyImageEventHandler). 

https://zh.docs.baslerweb.com/pylonapi/cpp/class_pylon_1_1_c_instant_camera#function-deregistercameraeventhandler

Also, they should all be set to nullptr (see cameraConfigurationEventHandler1) or none of them should be set to nullptr. I recommend not setting them, as they are automatically deleted upon deregistering, and only really used when they are being created.

Same applies to Hardware Trigger Configurations.

https://zh.docs.baslerweb.com/pylonapi/cpp/class_pylon_1_1_c_instant_camera#function-registerconfiguration

File: src/subwindows/serialSettingsDialog.cpp [hunk 2]
```diff
@@ -242,17 +248,17 @@ void SerialSettingsDialog::createForm() {
 
     QHBoxLayout *buttonsLayout = new QHBoxLayout();
 
-    connectButton = new QPushButton(tr("Connect"));
-    disconnectButton = new QPushButton(tr("Disconnect"));
+//    connectButton = new QPushButton(tr("Connect")); // moved to camera settings windows for clearer UX
```
Comment (sasfog) [inline src/subwindows/serialSettingsDialog.cpp:251]: If these are moved to another class, just delete them instead of commenting them out.

File: src/subwindows/singleCameraSettingsDialog.cpp [hunk 3]
```diff
@@ -583,63 +705,63 @@ SingleCameraSettingsDialog::~SingleCameraSettingsDialog() = default;
...
-    imageROIheightInputBox->setMaximum(singleCamera->getImageROIheightMax());
-    imageROIoffsetXInputBox->setMaximum(singleCamera->getImageROIwidthMax() -singleCamera->getImageROIwidth());
-    imageROIoffsetYInputBox->setMaximum(singleCamera->getImageROIheightMax() -singleCamera->getImageROIheight());
-
-    imageROIwidthMaxLabel->setText(QString("/ ") + QString::number(singleCamera->getImageROIwidthMax()));
-    imageROIheightMaxLabel->setText(QString("/ ") + QString::number(singleCamera->getImageROIheightMax()));
-    imageROIoffsetXMaxLabel->setText(QString("/ ") + QString::number(singleCamera->getImageROIwidthMax() -singleCamera->getImageROIwidth()));
-    imageROIoffsetYMaxLabel->setText(QString("/ ") + QString::number(singleCamera->getImageROIheightMax() -singleCamera->getImageROIheight()));
+    if(!camera->isOpen()) {
+        imageROIwidthInputBox->setMaximum(std::numeric_limits<short>::max());
```
Comment (sasfog) [inline src/subwindows/singleCameraSettingsDialog.cpp:739]: Why are these needed if the camera is not open?

File: src/subwindows/stereoCameraSettingsDialog.cpp [hunk 4]
```diff
@@ -759,18 +894,15 @@ void StereoCameraSettingsDialog::onSetImageROIoffsetY(int val) {
...
-    imageROIheightInputBox->setSingleStep(imageSizeChangeSingleStep/binningVal);
-    imageROIoffsetXInputBox->setMinimum(0);
-    imageROIoffsetXInputBox->setSingleStep(imageSizeChangeSingleStep/binningVal);
-    imageROIoffsetYInputBox->setMinimum(0);
-    imageROIoffsetYInputBox->setSingleStep(imageSizeChangeSingleStep/binningVal);
-}
-
 void StereoCameraSettingsDialog::updateImageROISettingsMax() {
+    if(!camera->isOpen()) {
+        imageROIwidthInputBox->setMaximum(std::numeric_limits<short>::max());
```
Comment (sasfog) [inline src/subwindows/stereoCameraSettingsDialog.cpp:899]: Why are these necessary?

File: src/subwindows/serialSettingsDialog.cpp (hunk 2 above)
Comment (kheki4) [reply src/subwindows/serialSettingsDialog.cpp:251]: Done

File: src/subwindows/singleCameraSettingsDialog.cpp (hunk 3 above)
Comment (kheki4) [reply src/subwindows/singleCameraSettingsDialog.cpp:739]: Agree, I removed it from the single camera version, but note that once we migrate single and stereo camera settings windows, this must be taken care of (the if(!camera->isOpen()... parts will be needed), as there the cameras can be opened only after the settings dialog is drawn, in a different step

File: src/subwindows/stereoCameraSettingsDialog.cpp (hunk 4 above)
Comment (kheki4) [reply src/subwindows/stereoCameraSettingsDialog.cpp:899]: The cameras can be opened only after the settings dialog is drawn, in a different step, and setting the queried max value (which is 0 in case of a closed camera) when the camera is closed could cause problems later, when upon opening the camera the actual values are to be filled there, before the real maximum is set. The max of short type is just a good arbitrary maximum when we do not know the real maximum

File: src/devices/stereoCamera.cpp (hunk 1 above)
Comment (kheki4) [reply src/devices/stereoCamera.cpp:1359]: All are now set to nullptr. Also added the same object management in single camera class, where software/hardware trigger configuration instances were not deregistered properly upon configuration change and camera closing.

**Review by sasfog — APPROVED** [review]
